flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/shuffle/FlinkKafkaShuffle.java

```diff
@@ -0,0 +1,248 @@
+ */
+@Experimental
+class FlinkKafkaShuffle {
+	static final String PRODUCER_PARALLELISM = "producer parallelism";
+	static final String PARTITION_NUMBER = "partition number";
+
+	/**
+	 * Writes to and reads from a kafka shuffle with the partition decided by keys.
+	 *
+	 * <p>Consumers should read partitions equal to the key group indices they are assigned.
+	 * The number of partitions is the maximum parallelism of the receiving operator.
+	 * This version only supports numberOfPartitions = consumerParallelism.
+	 *
+	 * @param inputStream Input stream to the kafka
+	 * @param topic Kafka topic
+	 * @param producerParallelism Parallelism of producer
+	 * @param numberOfPartitions Number of partitions
+	 * @param properties Kafka properties
+	 * @param fields Key positions from inputStream
+	 * @param <T> Input type
+	 */
+	public static <T> KeyedStream<T, Tuple> persistentKeyBy(
+			DataStream<T> inputStream,
+			String topic,
+			int producerParallelism,
+			int numberOfPartitions,
+			Properties properties,
+			int... fields) {
+		return persistentKeyBy(
+			inputStream,
+			topic,
+			producerParallelism,
+			numberOfPartitions,
+			properties,
+			keySelector(inputStream, fields));
+	}
+
+	/**
+	 * Writes to and reads from a kafka shuffle with the partition decided by keys.
+	 *
+	 * <p>Consumers should read partitions equal to the key group indices they are assigned.
+	 * The number of partitions is the maximum parallelism of the receiving operator.
+	 * This version only supports numberOfPartitions = consumerParallelism.
+	 *
+	 * @param inputStream Input stream to the kafka
+	 * @param topic Kafka topic
+	 * @param producerParallelism Parallelism of producer
+	 * @param numberOfPartitions Number of partitions
+	 * @param properties Kafka properties
+	 * @param keySelector key(K) based on inputStream(T)
+	 * @param <T> Input type
+	 * @param <K> Key type
+	 */
+	public static <T, K> KeyedStream<T, K> persistentKeyBy(
+			DataStream<T> inputStream,
+			String topic,
+			int producerParallelism,
+			int numberOfPartitions,
+			Properties properties,
+			KeySelector<T, K> keySelector) {
+		// KafkaProducer#propsToMap uses Properties purely as a HashMap without considering the default properties
+		// So we have to flatten the default property to first level elements.
+		Properties kafkaProperties = PropertiesUtil.flatten(properties);
+		kafkaProperties.setProperty(PRODUCER_PARALLELISM, String.valueOf(producerParallelism));
+		kafkaProperties.setProperty(PARTITION_NUMBER, String.valueOf(numberOfPartitions));
+
+		StreamExecutionEnvironment env = inputStream.getExecutionEnvironment();
+		TypeInformationSerializationSchema<T> schema =
+			new TypeInformationSerializationSchema<>(inputStream.getType(), env.getConfig());
+
+		writeKeyBy(inputStream, topic, kafkaProperties, keySelector);
+		return readKeyBy(topic, env, schema, kafkaProperties, keySelector);
+	}
+
+	/**
+	 * Writes to a kafka shuffle with the partition decided by keys.
```

Comment:
What does "Writes to a kafka shuffle" mean? I think you need to add more explanation.
This function also is only meaningful with `readKeyBy` so add a `@see`.